### DISCUSSION TITLE
Add new condition builder

### DISCUSSION THREAD

**Discussion by lukeholder** [craftcms/commerce#2290] [Ideas]
Add condition builder for logical conditions.

Will fix #181
Should fix #467 

Look into https://dabernathy89.github.io/vue-query-builder/ ?

Condition and effects should be separated for discounts and sales so effected products/variants can be different from conditions.

**Comment by lukeholder** [comment]
Instead of sales types or discount types, we should move to plugins providing condition objects we can use in a rules builder.

Conditions would really be like a fieldtype, but offer a way to match (true or false) an element

A condition object would provide:

inputHtml() like fields for the conditions params to be entered.
a list of elements the condition is compatible with
a match() method to match the condition with the element.

**Comment by lukeholder** [comment]
Need to also be able to generate all combinations of sale prices for all matching purchasables for each sale, so that the `hasSales` param on the product/variant/purchasable query can be a simple join.

**Comment by billmn** [comment]
Hi @lukeholder,
what about consider to use custom conditions also for payments?

This shoud fix this issue: https://github.com/craftcms/commerce/discussions/2467

Another example could be the need to disable a gateway based on the shipping zone.
